posts/2022/02/making-gpus-accessible-to-pydata-ecosystem-via-array-api.md

```diff
@@ -0,0 +1,486 @@
+     src="/images/2022/02/array_api_workflow.svg">
+    <i>Workflow for using Array API for writing code that works with multiple libraries</i>
+</p>
+
+#### Applying it on our demo example
+
+Now let's take a look at the code for our demo example (segmentation of the
+picture of greek coins in regions) to understand what it takes to make it run on GPU(s).
+
+```python
+# Author: Gael Varoquaux <gael.varoquaux@normalesup.org>, Brian Cheung
+# License: BSD 3 clause
+
+import time
+
+import numpy as np
+from scipy.ndimage.filters import gaussian_filter
+import matplotlib.pyplot as plt
+import skimage
+from skimage.data import coins
+from skimage.transform import rescale
+
+from sklearn.feature_extraction import image
+from sklearn.cluster import spectral_clustering
+from sklearn.utils.fixes import parse_version
+
+# these were introduced in skimage-0.14
+if parse_version(skimage.__version__) >= parse_version("0.14"):
+    rescale_params = {"anti_aliasing": False, "multichannel": False}
+else:
+    rescale_params = {}
+
+# load the coins as a numpy array
+orig_coins = coins()
+
+# Resize it to 20% of the original size to speed up the processing
+# Applying a Gaussian filter for smoothing prior to down-scaling
+# reduces aliasing artifacts.
+smoothened_coins = gaussian_filter(orig_coins, sigma=2)
+rescaled_coins = rescale(smoothened_coins, 0.2, mode="reflect", **rescale_params)
+
+# Convert the image into a graph with the value of the gradient on the
+# edges.
+graph = image.img_to_graph(rescaled_coins)
+
+# Take a decreasing function of the gradient: an exponential
+# The smaller beta is, the more independent the segmentation is of the
+# actual image. For beta=1, the segmentation is close to a voronoi
+beta = 10
+eps = 1e-6
+graph.data = np.exp(-beta * graph.data / graph.data.std()) + eps
+
+# Apply spectral clustering (this step goes much faster if you have pyamg
+# installed)
+N_REGIONS = 25
+
+# Plotting
+for assign_labels in ("kmeans", "discretize"):
+    t0 = time.time()
+    labels = spectral_clustering(
+        graph, n_clusters=N_REGIONS, assign_labels=assign_labels, random_state=42
+    )
+    t1 = time.time()
+    labels = labels.reshape(rescaled_coins.shape)
+
+    plt.figure(figsize=(5, 5))
+    plt.imshow(rescaled_coins, cmap=plt.cm.gray)
+    for l in range(N_REGIONS):
+        plt.contour(labels == l, colors=[plt.cm.nipy_spectral(l / float(N_REGIONS))])
+    plt.xticks(())
+    plt.yticks(())
+    title = "Spectral clustering: %s, %.2fs" % (assign_labels, (t1 - t0))
+    print(title)
+    plt.title(title)
+plt.show()
+```
+
+Let's walk through the code above to get a sense of what we're trying to achieve here.
+
+- We have a dataset of greek coins from pompeii (imported from `skimage.data.coins`)
+- We resize the image to 20% to speed up processing
+- Convert the image to graph data structure
+- Apply spectral clustering on the graph (via `kmeans` and `discretize` labelling).
+- Plot the resulting clustering
```

Comment:
I think it would be helpful to have the sections enumerated and apply the same enumeration to the comments in the code snippet above.
So `5. Plot the resulting clustering` and `# 5. Plotting`.